src/distributed_join.cuh

```diff
@@ -152,6 +163,162 @@ void all_to_all_comm(cudf::table_view input,
+    CUDA_RT_CALL(cudaStreamSynchronize(cudaStreamDefault));
+  }
+
+  template <typename T, std::enable_if_t<!std::is_integral<T>::value> * = nullptr>
+  void operator()(const void *uncompressed_data,
+                  size_t uncompressed_count,
+                  rmm::device_buffer &compressed_data,
+                  size_t &compressed_size)
+  {
+    assert(false && "Unsupported data type for cascaded compressor");
+  }
+};
+
+struct decompressor_functor {
+  template <typename T>
+  void operator()(const void *compressed_data,
+                  size_t compressed_size,
+                  void *output,
+                  size_t expected_output_count)
+  {
+    nvcomp::Decompressor<T> decompressor(compressed_data, compressed_size, cudaStreamDefault);
+    const size_t output_count = decompressor.get_num_elements();
+    assert(output_count == expected_output_count);
+
+    const size_t temp_size = decompressor.get_temp_size();
+    rmm::device_buffer temp_space(temp_size);
+
+    decompressor.decompress_async(
+      temp_space.data(), temp_size, static_cast<T *>(output), output_count, cudaStreamDefault);
+    CUDA_RT_CALL(cudaStreamSynchronize(cudaStreamDefault));
+  }
+};
+
+void all_to_all_comm_with_compression(cudf::table_view input,
+                                      vector<cudf::size_type> const &send_offset,
+                                      vector<rmm::device_buffer> &compressed_input,
+                                      vector<rmm::device_buffer> &compressed_output,
+                                      vector<vector<int64_t>> &compressed_recv_offset,
+                                      Communicator *communicator,
+                                      bool include_self = true)
+{
+  cudf::size_type ncols = input.num_columns();
+  compressed_input.resize(ncols);
+  compressed_output.resize(ncols);
+  compressed_recv_offset.resize(ncols);
+
+  int mpi_rank{communicator->mpi_rank};
+  int mpi_size{communicator->mpi_size};
+
+  for (cudf::size_type icol = 0; icol < ncols; icol++) {
+    cudf::column_view column   = input.column(icol);
+    cudf::data_type dtype      = column.type();
+    cudf::size_type dtype_size = cudf::size_of(dtype);
+
+    vector<rmm::device_buffer> compressed_column(mpi_size);
+    vector<size_t> compressed_column_size(mpi_size, 0);
+
+    // compress each partition in the column separately and store the result in *compressed_column*
+    for (int irank = 0; irank < mpi_size; irank++) {
+      if (!include_self && irank == mpi_rank) continue;
+
+      cudf::type_dispatcher(
+        dtype,
+        compression_functor{},
+        static_cast<const void *>(column.head<char>() + send_offset[irank] * dtype_size),
+        send_offset[irank + 1] - send_offset[irank],
+        compressed_column[irank],
+        compressed_column_size[irank]);
+    }
+
+    // calculate and communicate offsets for compressed column
+    vector<int64_t> compressed_send_offset(mpi_size + 1, -1);
```

Comment:
I agree we don't need to initialize it to anything. `compressed_send_offset` is designed to be filled later. I initialized it to `-1` because `-1` is not a valid index, so the code will immediately fail instead of carrying on quietly, if it is not filled.

Reply:
When it's initialized, I tried to figure out why, and anyone else that reads it will also have to figure out why.

If you don't initialize it then no one in the future will need to waste time trying to figure it out.  That might be good, for saving human time in the future.

Human time is worth more than compute time from failing immediately.

Reply:
Addressed in https://github.com/rapidsai/distributed-join/pull/43/commits/15d135347176f5103a5a6a80a57c2b88b41dbde0